### ISSUE TITLE
Workaround unavoidable marshaling on function pointers

### ISSUE THREAD

**Issue by PathogenDavid** [InfectedLibraries/Biohazrd#99]
Due to its heritage as a Windows technology, the CLR really wants to pretend `sizeof(bool) == 4` for the purposes of marshaling. Typically you can avoid this by using `MarshalAs(I1)`. The `bool` will still be marshaled to eliminate invalid boolean values (non-0/1 bools), but it'll be 1 byte at least. Unfortunately, it turns out the `calli` instruction (which is what function pointer calls are under the hood) perform marshaling too. (For some reason I was under the impression they didn't until I had a a "fun" time debugging bad marshaling in a case where a function returned a bool via `al` when the CLR used `eax`. This does make sense though since `calli` was used heavily for C++/CLI and is expected: https://github.com/dotnet/runtime/issues/40908)

In the case of char, the situation is *normally* better because if you mark it as being marshaled as `U2`, `U1`, or default marshaling on a Unicode PInvoke/struct it will become blittable. ([Source](https://github.com/dotnet/runtime/blob/29e9b5b7fd95231d9cd9d3ae351404e63cbb6d5a/src/coreclr/src/vm/fieldmarshaler.cpp#L233-L235)) You can enforce this behavior across the entire module using [`DefaultCharSetAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.defaultcharsetattribute?view=net-5.0). Unfortunately, again we can't use `MarshalAs` or similar on function pointers and `DefaultCharSetAttribute` does not apply to function pointers. (Whether this should be considered a bug is somewhat debatable since it'd make two modules with different default charsets incompatible for function pointer passing.)

In the long term, we'd like to see an escape hatch added to the CLR to allow these types to be safely used in unmanaged function pointers, and I'm working on a duo (now triplet) of proposals to improve the situation. Unfortunately we need a fix today.

As a (hopefully temporary) workaround, Biohazrd will synthesize two tiny wrapper structs to avoid marshaling.

One for booleans: ([SharpLab with JIT comparisons](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYmUo0GzNhwCSzRNz5iEkmfOLEA2iJgJ6UmABkIAT270YACntOXAaRaMAJtRiAI70cGoQggCUALqKAMzkCHAQftyMgg7ksFY25AByEKzicABC3NzCEIzEAN7E5A3kfAgs4oVwicmp6Zm+MOQAahGhCkT1jaYAsnAwABbcfhr8gq7TcwtLfIIA8nysaVDUAIIA5idJUFCtcFqCvr4nMeMNWAlUAGzkbFvSLP26iEK3AQ5GA5RWBSKpXByUYoMiz0aDQAvAA+cgAVUYUAgADMOEcoAAeSHXMoVWFIUHg1GuJK4+GjRFTGbzRbLVasjbLXb7bHHM4XK7FW73RiPWJEJGvZSfb53KR/cgAhBAkGk4rkyqMVxgirwxFItGY7F4gnEvWCKka6EUqq0+mM4gAXxMJDeeE+uHIABVZDAoLVETKPvlCmSYVVff7deCDVKkY1jcBRkjgx7Ppbo7BXDatbD44mk+iU4j07LKAAWcgABRazFj+uACITReNVAAnLrIqnGuXQ1hq3W+rnw5rI3Dm4bi8ou83ew1+57qfq/Tm8xPyBBrWPbdrC0XyMaIEfkaCma3KBmV4Js24s9ubwe2+iT8iz6XL0vM3G124N3acKPlmU6Xkar6nueZaXiGy5Zn+ja3o+AH7qBh5HhB75QUQrpEEAA))

```csharp
[StructLayout(LayoutKind.Sequential)]
public readonly struct NativeBoolean
{
    private readonly byte Value;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static implicit operator bool(NativeBoolean b)
        => Unsafe.As<NativeBoolean, bool>(ref b);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static implicit operator NativeBoolean(bool b)
        => Unsafe.As<bool, NativeBoolean>(ref b);

    // ...
}
```

And another for chars: ([SharpLab with JIT comparisons](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYmUo0GzNhwCSzRNz5iEkmfOLEA2iJgJ6UmABkIAT270YACntOXAaRaMAJtRiAI70cGoQggCUALqKAMzkCHAQftyMgg7ksFY25AByEKzicJwAFhAIxADexOR15HwILOKFcInJqemZ9FCl3Agw5ABqEaEKRLX1pgCycDB9fhr8gq6z89yLywDyfKxpUNQAggDmx0lQUM1wWoK+vscxk3VYCVQAbORsfLdSLIO6iEK/XIUnKCFcBSKJTBIMiT3qdQAvAA+cgAVUYUAgADMOIcoAAeSFXMoVJAgsHI1xJbGw8bwmZzBZLb6rJkbFmCHZ7TFHU7nS7FG53RgPWJEBEvZQfL4/P7kAEIIEIfKFElg1ygiqw+EIlHozE4vGErXIVVQ0kIKk0ukmCVTNbM5Zs9abb7clj7PlnWSC67pEVi+FS97kYDcbiCBV8QEwYGIxEQtXFS3kCDk4kpmHAOH2hFI1EQagjQShcgJsPF0Zwel58iM12cl1O927T28k4+i5XYWMe6POshvAfcOR6Ox4EAQkTmeh2vT5vV2pzuvq+qLJbLFeAVdLNfhDMdHOdR7dXLbXs7Ap7Ab7ooHktew7DEajiuV5cTprTGeTc5VK51nqhafpWm77hMdYNi2KynpyHqXvyvo3rcd5BoOT4jq+45KnGKrTpqMILrOqaAfma4gduu5jAeUFwSe7JnghHZId2Qq3v24qPtKL5ju+eGfkmFpEeS35keR5aFtRbTbrWCLQcerL0a2PIHFeyHsahnHBphvFvjGuFTjOf6pguYm5hJklptJoHAOMlnwgA9I55AAKIhBEUDknMUjUH5xAAL52kOHy4OQAAqsgwFA1Q6TxJEwpFsCEdqUgWRJ+pSPZ9RxaG35JW4CWpel5GZdldS5c+WAACzkAACk0zApSqaWrgWygAJyapE5XkJVoW1Q1viFSZMKtUBFGdd1vX9XpEVRUJS4qsRo3LiV+brrZcn1CFc0Fc1P4UmtbVWRAW20dxoajlG+1FctonZutwFpuddazdd83Jd+K3CcdE3tWdsmBcQQA===))

```chsarp
[StructLayout(LayoutKind.Sequential)]
public readonly struct NativeChar
{
    // For the purposes of sign extension the CLR treats char as unsigned.
    private readonly ushort Value;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static implicit operator char(NativeChar c)
        => Unsafe.As<NativeChar, char>(ref c);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static implicit operator NativeChar(char c)
        => Unsafe.As<char, NativeChar>(ref c);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator ==(NativeChar a, NativeChar b)
        => a.Value == b.Value;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator !=(NativeChar a, NativeChar b)
        => a.Value == b.Value;
    
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator ==(char a, NativeChar b)
        => a == b.Value;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator !=(char a, NativeChar b)
        => a == b.Value;
    
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator ==(NativeChar a, char b)
        => a.Value == b;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator !=(NativeChar a, char b)
        => a.Value == b;
        
    // Equals, etc...
}
```

One thing to note from the JIT disassembly is that for whatever reason the JIT does not enregister these structs for some reason. (It does enregister similar structs that wrap `int`, so I assume this is an issue of their size.) This probably causes a small perf hit with these structs, but it will pass them on the stack and in registers for P/Invokes. It seems like https://github.com/dotnet/runtime/pull/37745 (which fixed https://github.com/dotnet/runtime/issues/8016) *should've* caused these to be enregistered, but I need to look into it more.

It is unclear how valuable implementing the various instance methods on `System.Boolean` and `System.Char` on these types will be. In general native consumers should not use these types directly. but they can be desirable for making a native function calls where the return value is immediately consumed feel more natural. Initially I won't provide these, but they will likely be added in later.

Ideally we should not expose these types anywhere on the generated API surface except where absolutely necessary. Eventually we should support a few generation modes:

* Always use native wrappers
* Always use native wrappers at interop boundaries
* Only use when MarshalAs can't be used
* Always use if MarshalAs would've been required (IE: Everywhere that isn't a pointer to `bool` or `char`)
* Assume fictional/future modified runtime that doesn't marshal `bool` or `char` ever.

**Comment by PathogenDavid** [comment]
One problem that came up while implementing this: Using these types prevents using default parameter values.

Furthermore, it revealed a bug in the constant emit where it can emit a constant when it isn't valid to do so:

```csharp
public unsafe IIteratorLayer* addIterator(ITensor* tensor, int axis = 0, NativeBoolean reverse = 0)
```

**Comment by PathogenDavid** [comment]
Workaround implemented in 120b8de8aa6952950d21866c9fa79df74ac0df5f
